virtual_rainforest/core/registry.py

```diff
@@ -0,0 +1,147 @@
+
+    Args:
+        module_name: The name of the module containing the model to be registered
+
+    Raises:
+        ConfigurationError: If the module does not define a single Model class which is
+            a child of BaseModel.
+    """
+
+    # Try and import the module from the name to get a reference to the module
+    try:
+        module = import_module(module_name)
+    except ModuleNotFoundError as excep:
+        LOGGER.critical(f"Unknown module - registration failed: {module_name}")
+        raise excep
+
+    # Extract the last component of the module name to act as unique short name
+    module_name_short = module_name.split(".")[-1]
+    if module_name_short in MODULE_REGISTRY:
+        LOGGER.warning(f"Module already registered: {module_name_short}")
+        return
+
+    is_core = module_name == "virtual_rainforest.core"
+
+    LOGGER.info(f"Registering {module_name} module components")
+
+    # Check on the model argument.
+    if is_core and model is not None:
+        msg = "No model should be registered for the core module."
+        LOGGER.critical(msg)
+        raise RuntimeError(msg)
+    elif not is_core and model is None:
+        msg = "A model class is required to register model modules."
+        LOGGER.critical(msg)
+        raise RuntimeError(msg)
+    elif model is not None:
+        if module_name_short != model.model_name:
+            msg = f"The model_name attribute and module name differ in {module_name}"
+            LOGGER.critical(msg)
+            raise RuntimeError(msg)
+
+        # Register the resulting single model class
+        LOGGER.info(
+            f"Registering model class for {module_name_short} model: {model.__name__}"
+        )
+
+    # Register the schema
+    schema_file_name = "core_schema.json" if is_core else "model_schema.json"
+    with resources.as_file(
+        resources.files(module) / schema_file_name
+    ) as schema_file_path:
+        try:
+            schema = load_schema(
+                module_name=module_name_short, schema_file_path=schema_file_path
+            )
+        except Exception as excep:
+            LOGGER.critical(
+                f"Schema registration for {module_name_short} failed: check log"
+            )
+            raise excep
+
+    LOGGER.info(
+        "Schema registered for module %s: %s ", module_name_short, schema_file_path
+    )
+
+    # Find and register the constant dataclasses
+    constants_submodule = getattr(module, "constants", None)
+
+    if constants_submodule is None:
+        constants_classes = {}
+    else:
+        # Get all subclasses of ConstantsDataclass, excluding the ABC
+        constants_classes = {
+            class_name: class_obj
+            for class_name, class_obj in getmembers(constants_submodule)
+            if is_dataclass(class_obj)
+            and issubclass(class_obj, ConstantsDataclass)
+            and not class_obj.__name__ == "ConstantsDataclass"
+        }
```

Comment:
So, just to confirm, you not only want for it to be a subclass of `ConstantsClass`, it also needs to be defined as `dataclass` (since a subclass of a dataclass is not a dataclass), correct?

Reply:
The issue is that `issubclass` chokes if `class_obj` is not a class (and plenty of members aren't), so really all I need is a filter so that only classes get to that point. We could generically test for is a class, but since they are all dataclasses, this is a more usefully specific way to get to the same point. And then we don't want the ABC that has been imported to create the constants classes - I don't know if there is cleaner way than checking the `__name__`